src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/Subscriptions/DependenciesRuleChangeContext.cs

```diff
@@ -18,55 +18,36 @@ public DependenciesRuleChangeContext(ITargetFramework target, IProjectCatalogSna
         }
 
         public ITargetFramework ActiveTarget { get; }
+
         public IProjectCatalogSnapshot Catalogs { get; }
-        private ImmutableDictionary<ITargetFramework, IDependenciesChanges> _changes =
-            ImmutableDictionary.Create<ITargetFramework, IDependenciesChanges>();
 
-        public ImmutableDictionary<ITargetFramework, IDependenciesChanges> Changes
-        {
-            get
-            {
-                lock (_changesLock)
-                {
-                    return _changes;
-                }
-            }
-        }
+        public ImmutableDictionary<ITargetFramework, IDependenciesChanges> Changes { get; private set; } = ImmutableDictionary.Create<ITargetFramework, IDependenciesChanges>();
 
         public bool AnyChanges => Changes.Count != 0;
 
-        public void IncludeAddedChange(ITargetFramework targetFramework,
-                                       IDependencyModel ruleMetadata)
+        public void IncludeAddedChange(ITargetFramework targetFramework, IDependencyModel ruleMetadata)
         {
-            lock (_changesLock)
-            {
-                DependenciesChanges change = GetChanges(targetFramework);
-                change.ExcludeAddedChange(ruleMetadata);
-                change.IncludeAddedChange(ruleMetadata);
-            }
+            GetChanges(targetFramework).IncludeAddedChange(ruleMetadata);
         }
 
-        public void IncludeRemovedChange(ITargetFramework targetFramework,
-                                         IDependencyModel ruleMetadata)
+        public void IncludeRemovedChange(ITargetFramework targetFramework, IDependencyModel ruleMetadata)
         {
-            lock (_changesLock)
-            {
-                DependenciesChanges change = GetChanges(targetFramework);
-                change.ExcludeRemovedChange(ruleMetadata);
-                change.IncludeRemovedChange(ruleMetadata);
-            }
+            GetChanges(targetFramework).IncludeRemovedChange(ruleMetadata);
         }
 
         private DependenciesChanges GetChanges(ITargetFramework targetFramework)
         {
-            if (!_changes.TryGetValue(targetFramework, out IDependenciesChanges change))
+            lock (_changesLock)
             {
```

Comment:
Ditto, I would consider using ThreadingTools.ApplyChangeOptimistically to avoid the lock; it's perfect for this sort of pattern.

Reply:
Looking at `ApplyChangeOptimistically`, it could do with an overload that takes an operand to pass to the func, so you don't have to allocate a lambda when calling. I'll file a PR, if I can find the repo :)

A while back I put together set/list types that hid this internally behind normal collection interfaces/APIs. I'll dig the code out and push it to GitHub somewhere. Might be useful.

Reply:
Microsoft/vs-threading#408

Reply:
I found the code on an old laptop. It's from 2017, is completely undocumented and untested, but shows the idea. Pushed here: https://github.com/drewnoakes/optimistic-immutable-collections